rfcs/remote_channel.md

```diff
@@ -0,0 +1,650 @@
+   * Remove a handler for `postMessage` messages
+   */
+  removeEventListener(fn: (msg: Object) => void) {}
+
+ /**
+  * Wait for the next `postMessage` message and return it (after passing it to
+  * existing handlers)
+  */
+  async nextMessage(): Promise<Object> {}
+}
+
+class RemoteWindow {
+  /**
+   * Create a RemoteWindow. The dest parameter is either a
+   `SendChannel` object or the UUID for the channel. If ommitted a new
+   UUID is generated.
+   */
+  constructor(dest?: SendChannel | string) {}
+
+  /**
+   * Connect to the channel. Automatically called when sending the
+   * first message
+   */
+  async connect() {}
+
+  /**
+   * Close the channel and underlying websocket connections
+   */
+  async close() {}
+
+  /**
+   * Disconnect the RecvChannel running in the remote context, if any,
+   * on the server side
+   */
+   async pause() {}
+
+  /**
+   * Post the object msg to the remote, using JSON serialization
+   */
+   postMessage(msg: Object) {}
+
+  /**
+   * Run the function `fn` in the remote context, passing arguments
+   * `args`, and return the result after awaiting any returned
+   * promise.
+   *
+   * Arguments and return values are serialized as RemoteObjects.
+   */
+   async executeScript(fn: (args: ...any) => any, ..args: any): Promise<any> {}
+}
+
+/**
+ * Representation of a non-primitive type passed through a channel
+ */
+class RemoteObject {
+  type: string;
+  objectId: string | undefined;
+
+  /**
+   * Create a RemoteObject containing a handle to reference obj
+   */
+  static from(obj): RemoteObject
+
+  /**
+   * Return the local object referenced by the objectId of this RemoteObject,
+   * or null if there isn't a such an object in this realm.
+   */
+  toLocal(): Object?
+
+  /**
+   * Remove the objectId from the local cache. This means that future
+   * calls to `toLocal` with the same objectId will always return
+   * `null`.
+   */
+   delete()
+}
+
+/**
+ * Close all websockets in the current global that are being used for channels.
+ */
+function closeAllChannelSockets() {}
+
+///
+```
+
+### Resource Management
+
+With the stash-based approach, it's important to clean up the stash
+once no channels remain. Otherwise we end up leaking queues. The
+general approach is to explicitly store a refcount as part of the
+stash value. Each socket that connects increments the refcount, and
+decrements it once the socket is closed. Access to the stash is
+protected by a lock, so only one socket may touch the refcount at a
+time. In practice the fact that there can only be one `RecvChannel`
+per queue means that we can just count `SendChannel` sockets and
+enforce an at-most-one rule for `RecvChannel`. The underlying queue is
+deleted when no socket connections remain. This does mean that if we
+were to e.g. put a message on a queue, then delete all the channels
+before forwarding the message, and then use the UUID to create a new
+`RecvChannel` the queue would be deleted and the message lost. One can
```

Comment:
Well, I think the scheme you have here shouldn't leak more and more data as the tests run, so that's the most important thing. If the only scenario where the queue might be deleted too soon is when one deliberately reuses a UUID for a second `RecvChannel`, then that's not worth trying to fix.

Honestly I don't trust that I can see all the potential issues here, better to think about that when reviewing the implementation.